### PR TITLE
ENG-7677: Detailed DR catalog compatibility checks

### PR DESCRIPTION
**Pull request by afrogers**

### CODE REVIEW

File: src/catgen/in/javasrc/DRCatalogDiffEngine.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        }
+        return null;
+    }
+
+    @Override
+    protected String checkModifyWhitelist(final CatalogType suspect, final CatalogType prevType, final String field) {
+        if (suspect instanceof Table) {
+            if ("isdred".equalsIgnoreCase(field)) {
+                assert ((Boolean)suspect.getField("isDRed"));
+                return "Table " + suspect.getTypeName() + " has DR enabled on the master but not the replica";
```
Comment (nshi) [inline src/catgen/in/javasrc/DRCatalogDiffEngine.java]: What about the reverse, DR disabled on the master but enabled on the replica? Does it also end up in this branch?

Comment (nshi) [inline src/catgen/in/javasrc/DRCatalogDiffEngine.java]: Never mind, the catalog from the master only contains the DR tables. The case I mentioned earlier should be handled in the `checkAddDropWhitelist()`.

Comment (afrogers) [inline src/catgen/in/javasrc/DRCatalogDiffEngine.java]: Yep. Just what I was about to say.

On Tue, Apr 28, 2015 at 12:41 PM, Ning Shi <notifications@github.com> wrote:

> In src/catgen/in/javasrc/DRCatalogDiffEngine.java
> <https://github.com/VoltDB/voltdb/pull/2276#discussion_r29261913>:
>
> > +        if (ChangeType.ADDITION == changeType && suspect instanceof Table) {
> > +            assert ((Boolean)suspect.getField("isDRed"));
> > +            return "Missing DR table " + suspect.getTypeName() + " on replica cluster";
> > +        }
> > +        if (suspect instanceof Column || isUniqueIndex(suspect) || isUniqueIndexColumn(suspect) || isTableLimitDeleteStmt(suspect)) {
> > +            return "Missing " + suspect + " from " + suspect.getParent() + " on " + (ChangeType.ADDITION == changeType ? "replica" : "master");
> > +        }
> > +        return null;
> > +    }
> > +
> > +    @Override
> > +    protected String checkModifyWhitelist(final CatalogType suspect, final CatalogType prevType, final String field) {
> > +        if (suspect instanceof Table) {
> > +            if ("isdred".equalsIgnoreCase(field)) {
> > +                assert ((Boolean)suspect.getField("isDRed"));
> > +                return "Table " + suspect.getTypeName() + " has DR enabled on the master but not the replica";
>
> Never mind, the catalog from the master only contains the DR tables. The
> case I mentioned earlier should be handled in the checkAddDropWhitelist().
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/VoltDB/voltdb/pull/2276/files#r29261913>.
>



-- 

*[image: VoltDBLogoEmail]* <http://www.voltdb.com/>
Alex Rogers, Software Engineer
*VoltDB, Inc.*
*E: *arogers@voltdb.com

*voltdb.com* <http://www.voltdb.com/>
Follow VoltDB [image: twitter] <https://twitter.com/VoltDB> [image:
facebook] <https://www.facebook.com/VoltDB> [image: linkedin]
<https://www.linkedin.com/company/voltdb>

<http://www.linkedin.com/company/voltdb>

File: src/catgen/in/javasrc/DRCatalogDiffEngine.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            }
+            if ("estimatedtuplecount".equals(field)) {
+                return null;
+            }
+        } else if (suspect instanceof Column) {
+            if ("defaultvalue".equals(field) ||
+                "defaulttype".equals(field) ||
+                "matview".equals(field) ||
+                "aggregatetype".equals(field) ||
+                "matviewsource".equals(field)) {
```
Comment (nshi) [inline src/catgen/in/javasrc/DRCatalogDiffEngine.java]: Will checking the blacklist be easier here? I think `nullable` should also be in the whitelist.

What are the `matview`, `aggregatetype`, and `matviewsource` fields?

Comment (afrogers) [inline src/catgen/in/javasrc/DRCatalogDiffEngine.java]: Checking a whitelist seemed safer -- especially if we're only going to
warn, it struck me as better to be a bit overaggressive. 'matview',
'aggregatetype', and 'matviewsource' pertain only to materialized views, I
believe. Maybe better to just assert these *aren't* the field, but this
seemed fine.

Also, nullable was intentionally excluded -- think about the case where
it's not nullable on the replica, but nullable on the master (and the
inverse for failover/failback).

On Tue, Apr 28, 2015 at 12:45 PM, Ning Shi <notifications@github.com> wrote:

> In src/catgen/in/javasrc/DRCatalogDiffEngine.java
> <https://github.com/VoltDB/voltdb/pull/2276#discussion_r29262369>:
>
> > +    @Override
> > +    protected String checkModifyWhitelist(final CatalogType suspect, final CatalogType prevType, final String field) {
> > +        if (suspect instanceof Table) {
> > +            if ("isdred".equalsIgnoreCase(field)) {
> > +                assert ((Boolean)suspect.getField("isDRed"));
> > +                return "Table " + suspect.getTypeName() + " has DR enabled on the master but not the replica";
> > +            }
> > +            if ("estimatedtuplecount".equals(field)) {
> > +                return null;
> > +            }
> > +        } else if (suspect instanceof Column) {
> > +            if ("defaultvalue".equals(field) ||
> > +                "defaulttype".equals(field) ||
> > +                "matview".equals(field) ||
> > +                "aggregatetype".equals(field) ||
> > +                "matviewsource".equals(field)) {
>
> Will checking the blacklist be easier here? I think nullable should also
> be in the whitelist.
>
> What are the matview, aggregatetype, and matviewsource fields?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/VoltDB/voltdb/pull/2276/files#r29262369>.
>



-- 

*[image: VoltDBLogoEmail]* <http://www.voltdb.com/>
Alex Rogers, Software Engineer
*VoltDB, Inc.*
*E: *arogers@voltdb.com

*voltdb.com* <http://www.voltdb.com/>
Follow VoltDB [image: twitter] <https://twitter.com/VoltDB> [image:
facebook] <https://www.facebook.com/VoltDB> [image: linkedin]
<https://www.linkedin.com/company/voltdb>

<http://www.linkedin.com/company/voltdb>

File: src/catgen/in/javasrc/DRCatalogDiffEngine.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            if ("cost".equals(field) ||
+                "seqscancount".equals(field) ||
+                "explainplan".equals(field) ||
+                "indexesused".equals(field) ||
+                "cachekeyprefix".equals(field)) {
+                return null;
+            }
+        } else if (isUniqueIndex(suspect) == isUniqueIndex(prevType) && !isUniqueIndexColumn(suspect)) {
+            return null;
+        }
```
Comment (nshi) [inline src/catgen/in/javasrc/DRCatalogDiffEngine.java]: What about constraints?

Comment (afrogers) [inline src/catgen/in/javasrc/DRCatalogDiffEngine.java]: I could not find any cases we cared about on the constraint object itself.
Every case I could see was handled on the Index object itself.

On Tue, Apr 28, 2015 at 1:01 PM, Ning Shi <notifications@github.com> wrote:

> In src/catgen/in/javasrc/DRCatalogDiffEngine.java
> <https://github.com/VoltDB/voltdb/pull/2276#discussion_r29263952>:
>
> > +                "matview".equals(field) ||
> > +                "aggregatetype".equals(field) ||
> > +                "matviewsource".equals(field)) {
> > +                return null;
> > +            }
> > +        } else if (isTableLimitDeleteStmt(suspect)) {
> > +            if ("cost".equals(field) ||
> > +                "seqscancount".equals(field) ||
> > +                "explainplan".equals(field) ||
> > +                "indexesused".equals(field) ||
> > +                "cachekeyprefix".equals(field)) {
> > +                return null;
> > +            }
> > +        } else if (isUniqueIndex(suspect) == isUniqueIndex(prevType) && !isUniqueIndexColumn(suspect)) {
> > +            return null;
> > +        }
>
> What about constraints?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/VoltDB/voltdb/pull/2276/files#r29263952>.
>



-- 

*[image: VoltDBLogoEmail]* <http://www.voltdb.com/>
Alex Rogers, Software Engineer
*VoltDB, Inc.*
*E: *arogers@voltdb.com

*voltdb.com* <http://www.voltdb.com/>
Follow VoltDB [image: twitter] <https://twitter.com/VoltDB> [image:
facebook] <https://www.facebook.com/VoltDB> [image: linkedin]
<https://www.linkedin.com/company/voltdb>

<http://www.linkedin.com/company/voltdb>

Comment (afrogers) [inline src/catgen/in/javasrc/DRCatalogDiffEngine.java]: See the comment on line 359 in CatalogDiffEngine for a good explanation.

On Tue, Apr 28, 2015 at 1:07 PM, Alex Rogers <arogers@voltdb.com> wrote:

> I could not find any cases we cared about on the constraint object itself.
> Every case I could see was handled on the Index object itself.
>
> On Tue, Apr 28, 2015 at 1:01 PM, Ning Shi <notifications@github.com>
> wrote:
>
>> In src/catgen/in/javasrc/DRCatalogDiffEngine.java
>> <https://github.com/VoltDB/voltdb/pull/2276#discussion_r29263952>:
>>
>> > +                "matview".equals(field) ||
>> > +                "aggregatetype".equals(field) ||
>> > +                "matviewsource".equals(field)) {
>> > +                return null;
>> > +            }
>> > +        } else if (isTableLimitDeleteStmt(suspect)) {
>> > +            if ("cost".equals(field) ||
>> > +                "seqscancount".equals(field) ||
>> > +                "explainplan".equals(field) ||
>> > +                "indexesused".equals(field) ||
>> > +                "cachekeyprefix".equals(field)) {
>> > +                return null;
>> > +            }
>> > +        } else if (isUniqueIndex(suspect) == isUniqueIndex(prevType) && !isUniqueIndexColumn(suspect)) {
>> > +            return null;
>> > +        }
>>
>> What about constraints?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/VoltDB/voltdb/pull/2276/files#r29263952>.
>>
>
>
>
> --
>
> *[image: VoltDBLogoEmail]* <http://www.voltdb.com/>
> Alex Rogers, Software Engineer
> *VoltDB, Inc.*
> *E: *arogers@voltdb.com
>
> *voltdb.com* <http://www.voltdb.com/>
> Follow VoltDB [image: twitter] <https://twitter.com/VoltDB> [image:
> facebook] <https://www.facebook.com/VoltDB> [image: linkedin]
> <https://www.linkedin.com/company/voltdb>
>
> <http://www.linkedin.com/company/voltdb>
>



-- 

*[image: VoltDBLogoEmail]* <http://www.voltdb.com/>
Alex Rogers, Software Engineer
*VoltDB, Inc.*
*E: *arogers@voltdb.com

*voltdb.com* <http://www.voltdb.com/>
Follow VoltDB [image: twitter] <https://twitter.com/VoltDB> [image:
facebook] <https://www.facebook.com/VoltDB> [image: linkedin]
<https://www.linkedin.com/company/voltdb>

<http://www.linkedin.com/company/voltdb>

Comment (nshi) [inline src/catgen/in/javasrc/DRCatalogDiffEngine.java]: > See the comment on line 359 in CatalogDiffEngine for a good explanation.

Ah, ok, makes sense.

File: tests/frontend/org/voltdb/catalog/TestDRCatalogDiffs.java [hunk 4]
```diff
@@ -0,0 +1,669 @@
...
+                "CREATE TABLE T1 (C1 INTEGER NOT NULL, C2 INTEGER NOT NULL);\n" +
+                "CREATE TABLE T2 (C1 INTEGER NOT NULL, C2 INTEGER NOT NULL);\n" +
+                "DR TABLE T1;\n" +
+                "DR TABLE T2;";
+        CatalogDiffEngine diff = runCatalogDiff(masterSchema, replicaSchema);
+        assertTrue(diff.errors(), diff.supported());
+    }
+
+    @Test
+    public void testExtraDRTableOnReplica() throws Exception {
```
Comment (nshi) [inline tests/frontend/org/voltdb/catalog/TestDRCatalogDiffs.java]: This is the same as `testMissingDRTableOnMaster`?

File: tests/frontend/org/voltdb/catalog/TestDRCatalogDiffs.java [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+                "PARTITION TABLE T1 ON COLUMN C1;\n" +
+                "DR TABLE T1;";
+
+        CatalogDiffEngine diff = runCatalogDiff(masterSchema, replicaSchema);
+        assertFalse(diff.supported());
+        assertTrue(diff.errors().contains("Missing Statement{limit_delete} from Table{T1} on replica"));
+    }
+
+    @Test
+    public void testDifferentDeletePolicy() throws Exception {
```
Comment (nshi) [inline tests/frontend/org/voltdb/catalog/TestDRCatalogDiffs.java]: Maybe another test that has the same DELETE clause but with a different constant?

**Comment by nshi** [comment]
Great set of unit tests!

Who handles the case there there is no DR table on the master or replica side? Can the diff engine handle that? It's basically an empty catalog string, right?

**Comment by afrogers** [comment]
Should be an empty catalog string. I'll test it.

On Tue, Apr 28, 2015 at 1:32 PM, Ning Shi <notifications@github.com> wrote:

> Great set of unit tests!
>
> Who handles the case there there is no DR table on the master or replica
> side? Can the diff engine handle that? It's basically an empty catalog
> string, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/VoltDB/voltdb/pull/2276#issuecomment-97145022>.
>



-- 

*[image: VoltDBLogoEmail]* <http://www.voltdb.com/>
Alex Rogers, Software Engineer
*VoltDB, Inc.*
*E: *arogers@voltdb.com

*voltdb.com* <http://www.voltdb.com/>
Follow VoltDB [image: twitter] <https://twitter.com/VoltDB> [image:
facebook] <https://www.facebook.com/VoltDB> [image: linkedin]
<https://www.linkedin.com/company/voltdb>

<http://www.linkedin.com/company/voltdb>

File: tests/frontend/org/voltdb/catalog/TestDRCatalogDiffs.java (hunk 4 above)
Comment (afrogers) [inline tests/frontend/org/voltdb/catalog/TestDRCatalogDiffs.java]: So it is...

On Tue, Apr 28, 2015 at 1:20 PM, Ning Shi <notifications@github.com> wrote:

> In tests/frontend/org/voltdb/catalog/TestDRCatalogDiffs.java
> <https://github.com/VoltDB/voltdb/pull/2276#discussion_r29265670>:
>
> > +    public void testTableNotDROnMaster() throws Exception {
> > +        String masterSchema =
> > +                "CREATE TABLE T1 (C1 INTEGER NOT NULL, C2 INTEGER NOT NULL);\n" +
> > +                "CREATE TABLE T2 (C1 INTEGER NOT NULL, C2 INTEGER NOT NULL);\n" +
> > +                "DR TABLE T2;";
> > +        String replicaSchema =
> > +                "CREATE TABLE T1 (C1 INTEGER NOT NULL, C2 INTEGER NOT NULL);\n" +
> > +                "CREATE TABLE T2 (C1 INTEGER NOT NULL, C2 INTEGER NOT NULL);\n" +
> > +                "DR TABLE T1;\n" +
> > +                "DR TABLE T2;";
> > +        CatalogDiffEngine diff = runCatalogDiff(masterSchema, replicaSchema);
> > +        assertTrue(diff.errors(), diff.supported());
> > +    }
> > +
> > +    @Test
> > +    public void testExtraDRTableOnReplica() throws Exception {
>
> This is the same as testMissingDRTableOnMaster?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/VoltDB/voltdb/pull/2276/files#r29265670>.
>



-- 

*[image: VoltDBLogoEmail]* <http://www.voltdb.com/>
Alex Rogers, Software Engineer
*VoltDB, Inc.*
*E: *arogers@voltdb.com

*voltdb.com* <http://www.voltdb.com/>
Follow VoltDB [image: twitter] <https://twitter.com/VoltDB> [image:
facebook] <https://www.facebook.com/VoltDB> [image: linkedin]
<https://www.linkedin.com/company/voltdb>

<http://www.linkedin.com/company/voltdb>

**Comment by nshi** [comment]
Squash merged externally.

**Pull request closed without merging**